src/solvers/unsteady_solver.C

```diff
@@ -281,6 +294,168 @@ void UnsteadySolver::integrate_adjoint_sensitivity(const QoISet & qois, const Pa
+      }
+    }
+  }
+
+  // Advance to t_j+1
+  _system.time = _system.time + _system.deltat;
+
+  // Get t_j+1
+  Real time_right = _system.time;
+
+  // We will need to use the last step deltat for the weighted residual evaluation
+  last_step_deltat = _system.deltat;
+
+  // The adjoint error estimate expression for a backwards facing step
+  // scheme needs the adjoint for the last time instant, so save the current adjoint for future use
+  NumericVector<Number> & current_adjoint = _system.get_adjoint_solution(0);
+  old_adjoint = current_adjoint.clone();
+
+  // Retrieve the state and adjoint vectors for the next time instant
+  retrieve_timestep();
+
+  // Swap for residual weighting
+  current_adjoint.swap(*old_adjoint);
+
+  // Swap out the deltats as we did for the left side
+  next_step_deltat = _system.deltat;
+  _system.deltat = last_step_deltat;
+
+  // Get f(t_j+1)
+  ErrorVector QoI_elementwise_error_right;
+  std::map<int, Real> global_spatial_errors_right;
+
+  _system.update();
+
+  // The residual has to be evaluated at the last time
+  _system.time = _system.time - last_step_deltat;
+
+  adjoint_refinement_error_estimator.estimate_error(_system, QoI_elementwise_error_right);
+
+  // Shift the time back
+  _system.time = _system.time + last_step_deltat;
+
+  // Set the system deltat back to what it needs to be able to march to the next time
+  _system.deltat = next_step_deltat;
+
+  // Swap back now that the residual weighting is done
+  current_adjoint.swap(*old_adjoint);
+
+  // Also get the right side contributions for the spatially integrated errors for all the QoIs in the QoI set
+  for (auto j : make_range(_system.n_qois()))
+  {
+    // Skip this QoI if not in the QoI Set
+    if (adjoint_refinement_error_estimator.qoi_set().has_index(j))
+    {
+      qoi_error_estimates_right[j] = adjoint_refinement_error_estimator.get_global_QoI_error_estimate(j);
+    }
+  }
+
+  // Error contribution from this timestep
+  for(unsigned int i = 0; i < QoI_elementwise_error.size(); i++)
+    QoI_elementwise_error[i] = ((QoI_elementwise_error_right[i] + QoI_elementwise_error_left[i])/2.)*(time_right - time_left);
+
+  // QoI set spatially integrated errors contribution from this timestep
+  for (auto j : make_range(_system.n_qois()))
+  {
+    // Skip this QoI if not in the QoI Set
+    if (adjoint_refinement_error_estimator.qoi_set().has_index(j))
+    {
+      if(QoI_time_instant[j] == NULL)
+      {
+        (_system.qoi_error_estimates)[j] = ((qoi_error_estimates_left[j] + qoi_error_estimates_right[j])/2.)*(time_right - time_left);
+      }
+      else if(time_right <= *(QoI_time_instant[j]) + TOLERANCE)
+      {
+        (_system.qoi_error_estimates)[j] = qoi_error_estimates_right[j]*last_step_deltat;
+      }
+      else
+      {
+        (_system.qoi_error_estimates)[j] = 0.0;
+        libmesh_warning("Error estimate contribution for a time instant past the instantaneous QoI evaluation time has been set to zero.");
```

Comment:
When do we expect this to happen?  Should this be an error rather than a warning?

Reply:
QoI0 is instantaneous and QoI2 is smooth and has support past the time instant.

Reply:
QThat seems awfully specific and awfully cryptic.  I assume you mean that, because one smooth QoI exists, the user will be calling this function at every time step, and because an instantaneous QoI also exists, the library will notice that it's being asked for error terms coming from time steps after that QoI instant?  That's not even worth a warning, then, except in the case where the QoI set doesn't have any other QoIs.

Speaking of awfully cryptic - I just noticed now that QoI_time_instant isn't even documented anywhere yet.  Right now the user base of this code is basically "you", and maybe I have a fighting chance because I can go back through IM history to figure out what various things are for, but I'm pretty sure if I showed these APIs to another core developer, much less a typical user, they'd have no idea where to begin.  Hopefully the new example will help with that, but examples are a good addition to documentation, not a replacement for it.

Reply:
Added documentation for QoI_time_instant in time_solver.h with the comments for the base integrate_adjoint_refinement_error_estimate function declaration.